cloudinit/config/cc_mounts.py

```diff
@@ -367,180 +373,217 @@ def handle(name: str, cfg: Config, cloud: Cloud, args: list) -> None:
-            )
+        if not isinstance(line, list):
+            LOG.warning("Mount option not a list, ignoring: %s", line)
             continue
 
-        start = str(cfgmnt[i][0])
-        sanitized = sanitize_devname(
+        start = str(line[0])
+        sanitized_devname = sanitize_devname(
             start, cloud.device_name_to_device, aliases=device_aliases
         )
-        if sanitized != start:
-            LOG.debug("changed %s => %s", start, sanitized)
+        if sanitized_devname_is_valid(start, sanitized_devname, fstab_devs):
+            updated_line = [sanitized_devname] + line[1:]
+        else:
+            updated_line = line
 
-        if sanitized is None:
-            LOG.debug("Ignoring nonexistent named mount %s", start)
-            continue
-        elif sanitized in fstab_devs:
-            LOG.info(
-                "Device %s already defined in fstab: %s",
-                sanitized,
-                fstab_devs[sanitized],
-            )
-            continue
+        # Ensure all tokens are strings as users may not have quoted them
+        updated_line = [
+            str(token) for token in updated_line if token is not None
+        ]
 
-        cfgmnt[i][0] = sanitized
+        # fill in values with defaults from defvals above
+        updated_line += default_fields[len(updated_line) :]
 
-        # in case the user did not quote a field (likely fs-freq, fs_passno)
-        # but do not convert None to 'None' (LP: #898365)
-        for j in range(len(cfgmnt[i])):
-            if cfgmnt[i][j] is None:
-                continue
-            else:
-                cfgmnt[i][j] = str(cfgmnt[i][j])
+        updated_lines.append(updated_line)
+    return updated_lines
 
-    for i in range(len(cfgmnt)):
-        # fill in values with defaults from defvals above
-        for j in range(len(defvals)):
-            if len(cfgmnt[i]) <= j:
-                cfgmnt[i].append(defvals[j])
-            elif cfgmnt[i][j] is None:
-                cfgmnt[i][j] = defvals[j]
-
-        # if the second entry in the list is 'None' this
-        # clears all previous entries of that same 'fs_spec'
-        # (fs_spec is the first field in /etc/fstab, ie, that device)
-        if cfgmnt[i][1] is None:
-            for j in range(i):
-                if cfgmnt[j][0] == cfgmnt[i][0]:
-                    cfgmnt[j][1] = None
-
-    # for each of the "default" mounts, add them only if no other
-    # entry has the same device name
-    for defmnt in defmnts:
-        start = defmnt[0]
+
+def remove_nonexistent_devices(mounts: List[List[str]]) -> List[List[str]]:
+    """Remove any entries that have a device name that doesn't exist.
+
+    If the second field of a mount line is None (not the string, the value),
+    we skip it along with any other entries that came before it that share
+    the same device name.
+    """
+    actlist = []
+    dev_denylist = []
+    for line in mounts[::-1]:
```

Comment:
Because we have some weird rules that would allow a mount declaration like this:
```
/dev/sda1 /mnt1
/dev/sda1 None
/dev/sda1 /mnt2
```

The `None` on line 2 means "disregard this line and any /dev/sda1 entries that came before it". The first reversal is to make it easy to know which lines to throw out that we come upon. If we previously saw a None having the same first entry, we know to throw it out.

Since we accumulated our act list in reverse order, the last reverse is to put the list back into it's original order (after we've removed the lines we want to remove).